dali/pipeline/graph_descr.cc

```diff
@@ -71,358 +73,444 @@ void CheckOpConstraints(const OpSpec &spec) {
+        op_nodes_ref[oid].parents.insert(new_id);
+      }
+    };
+    remove_id_in_family_op(left, left_id);
+    remove_id_in_family_op(right, right_id);
+    add_id_in_family_op(left, right_id);
+    add_id_in_family_op(right, left_id);
   }
 
-  // Remove this nodes entry from the id map. This will
-  // effectively decrement all node ids after this node
-  // to fill the gap.
-  //
-  auto type_and_idx = id_to_node_map_[id];
-  DALIOpType type = type_and_idx.first;
-  int idx = type_and_idx.second;
-  id_to_node_map_.erase(id_to_node_map_.begin() + id);
-
-  // Remove the typed node object for the target node.
-  // We will then need to update the id map entry for
-  // all nodes of this type that follow the deleted node
-  switch (type) {
-  case DALI_CPU:
-    cpu_nodes_.erase(cpu_nodes_.begin() + idx);
-
-    for (size_t i = idx; i < cpu_nodes_.size(); ++i) {
-      OpNode &cpu_node = this->cpu_node(i);
-      id_to_node_map_[cpu_node.id].second = i;
-    }
-    break;
-  case DALI_GPU:
-    gpu_nodes_.erase(gpu_nodes_.begin() + idx);
+  // Swap the nodes
+  left.id = right_id;
+  right.id = left_id;
+  std::swap(left, right);
+}
 
-    for (size_t i = idx; i < gpu_nodes_.size(); ++i) {
-      OpNode &gpu_node = this->gpu_node(i);
-      id_to_node_map_[gpu_node.id].second = i;
-    }
-    break;
-  case DALI_MIXED:
-    mixed_nodes_.erase(mixed_nodes_.begin() + idx);
+void OpGraph::RemoveOpNode(OpNodeId id) {
+  DALI_ENFORCE_VALID_INDEX(id, (Index)op_nodes_.size());
+  auto &target_op = op_nodes_[id];
+  DALI_ENFORCE(target_op.children.empty(), "Overwritten ops cannot have any children.");
+  DALI_ENFORCE(target_op.children_tensors.empty(),
+               "All produced tensors should be removed before removing op"
+               " and list of children tensors should be invalidated.");
+  for (OpNodeId i = id + 1; i < static_cast<int>(op_nodes_.size()); i++) {
+    // Move from i to i - 1
+    SwapOpNodes(i - 1, i);
+  }
+  // Remove the edge from parent Ops
+  for (auto parent_id : op_nodes_.back().parents) {
+    Node(parent_id).children.erase(op_nodes_.back().id);
+  }
+  // assume that we removed one element
+  op_nodes_.pop_back();
+}
 
-    for (size_t i = idx; i < mixed_nodes_.size(); ++i) {
-      OpNode &mixed_node = this->mixed_node(i);
-      id_to_node_map_[mixed_node.id].second = i;
-    }
-    break;
-  case DALI_SUPPORT:
-    support_nodes_.erase(support_nodes_.begin() + idx);
+namespace {
+
+template <typename T>
+void RemoveVectorElement(T& vector, int index) {
+  std::swap(vector[index], vector.back());
+  vector.pop_back();
+}
+
+}  // namespace
+
+// Op Removal Process:
+// 1. Validate we can remove it (it has no children & no consumers for produced tensors)
+// 2. Remove tensors it produces
+// 3. Remove us from consumer lists of parent tensors
+// 4. Remove the OpNode with edges from parent Ops
+// 6. Correct the partitions as we changed the ids while removing OpNode
+void OpGraph::RemoveOp(OpNodeId id) {
```

Comment:
Please add unit test for this function

Reply:
https://github.com/NVIDIA/DALI/blob/7c7249e2b4ab3ff04754883e6f911f166b35580e/dali/pipeline/op_graph_test.cc#L288

There already are unit tests for OpGraph